### PR TITLE
Upgrade to use multiple memory map files

### PR DESCRIPTION
**Pull request by MichaelKim20**
The file is divided into multi parts.
The number of blocks in a file is fixed.
Store the position of each block using an array.
Replace with an index file later.

MMFile-BlockStorage-Index

### CODE REVIEW

File: source/agora/common/BlockStorage.d [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    /// Ctor
+    public this (string path)
+    {
+        this.is_mapped = false;
+        this.path = path;
+    }
+
+    public ~this ()
+    {
+        this.unMap();
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d]: You cannot use the GC from a destructor.
Since D has a garbage collector, when a reference is collected by the GC, you have no guarantee that the GC-allocated references it holds are still alive.
If you had this guarantee, it would be impossible to finalize objects with cyclic references.

Comment (MichaelKim20) [reply source/agora/common/BlockStorage.d]: Okay. I'll remove this part.

File: source/agora/common/BlockStorage.d [hunk 2]
```diff
@@ -62,7 +71,40 @@ public class BlockStorage
...
+            length = The size of data
+
+    ***************************************************************************/
+
+    private void writeDataLength (size_t length)
+    {
+        ubyte[] values;
+        values.length = size_t.sizeof;
+        *cast(size_t*)&(values[0]) = length;
+        foreach (idx, ref e; values)
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d]:
```suggestion
        foreach (idx, e; (cast(const ubyte*)&length)[0 .. size_t.sizeof])
```

**Comment by Geod24** [comment]
I think it would be great to break this PR done into smaller commits.

**Comment by Geod24** [comment]
It's pretty hard to follow, as the commits are not atomic. For example the first 2 commits override each other quite a lot.
Style wise, use `private immutable ConstantName = 42;` instead of `const CONSTANT_NAME = 42;`.

File: source/agora/common/BlockStorage.d [hunk 3]
```diff
@@ -159,35 +298,346 @@ public class BlockStorage
...
+
+        Read type of `size_t` data
+
+        Params:
+            pos = position of memory mapped file
+            value = type of `size_t`
+
+    ***************************************************************************/
+
+    private void writeSizeT (size_t pos, size_t value) @trusted
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:423]: W.r.t. `readDataLength`, `writeDataLength`, `writeSizeT`, `readSizeT`: This is what the serialization / deserialization module is for.

File: source/agora/common/BlockStorage.d [hunk 4]
```diff
@@ -37,56 +73,125 @@ public class BlockStorage
...
+                    MmFile.Mode.readWrite,
+                    resize + MFILE_HEAD_SIZE + MFILE_RESERVE_SIZE,
+                    null
+                );
+            this.writeDataLength(resize);
+            this.data_size = resize;
+        }
+        else
+        {
+            if (name.exists())
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:171]: Can be turned into an `else if`

File: source/agora/common/BlockStorage.d [hunk 5]
```diff
@@ -159,35 +298,346 @@ public class BlockStorage
...
+        size_t height;
+        ubyte[HASH_SIZE] hash;
+        size_t pos;
+        string file_name = buildPath(this.path, "index.dat");
+
+        this.height_idx.clear();
+        this.hash_idx.clear();
+
+        if (file_name.exists)
+        {
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:563]: Use early returns to reduce indentation.
In this case, instead of wrapping the rest of the function in a `if`, use:
```d
if (!file_name.exists())
    return;
```

This way the code is simpler to read.

File: source/agora/common/BlockStorage.d [hunk 6]
```diff
@@ -159,35 +298,346 @@ public class BlockStorage
...
+        Params:
+            block = `Block` to read
+            hash = `Hash` of `Block`
+
+        Returns:
+            Returns true if success, otherwise returns false.
+
+    ***************************************************************************/
+
+    public bool readBlock (ref Block block, Hash hash) @safe
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:354]: The whole point of memory mapping is to return a `const` pointer, so that there is only ever one copy of the block. This copy the block.

File: source/agora/common/BlockStorage.d [hunk 7]
```diff
@@ -118,37 +234,60 @@ public class BlockStorage
...
 
-    public bool saveBlock (const ref Block block)
+    public bool saveBlock (const ref Block block) @safe
     {
-        if (block.header.height != this.block_positions.length)
+        if ((this.height_idx.length > 0) &&
+            (this.height_idx.back.height >= block.header.height))
             return false;
 
         const ubyte[] serialized_block = serializeFull(block);
```
Comment (Geod24) [inline source/agora/common/BlockStorage.d:243]: Instead of allocating memory to serialize, it should be done straight to the file.

**Comment by Geod24** [comment]
Also can you rebase against `v0.x.x` so we get coverage report ?

**Review by Geod24 — APPROVED** [review]
Actually on a second take, this does not break anything, so let's move forward and fix as needed.
